plutus-ledger-api/test-plugin/Spec/MintValue/V3.hs

```diff
@@ -0,0 +1,153 @@
+-- MintValue properties --------------------------------------------------------
+
+prop_MintValueBuiltinData :: Either Value MintValue -> Bool
+prop_MintValueBuiltinData values =
+  let (value, mintValue) =
+        case values of
+          Left v   -> (v, coerce v)
+          Right mv -> (coerce mv, mv)
+   in toBuiltinData mintValue == toBuiltinData value
+
+prop_AssetClassIsEitherMintedOrBurned :: MintValue -> Bool
+prop_AssetClassIsEitherMintedOrBurned (mint :: MintValue) =
+  let assetClasses v = [AssetClass (cs, tn) | (cs, tn, _q) <- flattenValue v]
+      minted = assetClasses (mintValueMinted mint)
+      burned = assetClasses (mintValueBurned mint)
+   in List.all (`List.notElem` burned) minted
+
+prop_MintValueMintedIsPositive :: MintValue -> Bool
+prop_MintValueMintedIsPositive mint =
+  List.all
+    (List.all (> 0) . Map.elems)
+    (Map.elems (getValue (mintValueMinted mint)))
+
+prop_MintValueBurnedIsPositive :: MintValue -> Bool
+prop_MintValueBurnedIsPositive mint =
+  List.all
+    (List.all (> 0) . Map.elems)
+    (Map.elems (getValue (mintValueBurned mint)))
+
+--------------------------------------------------------------------------------
+-- Haskell-evaluated properties ------------------------------------------------
+
+testPropsInHaskell :: TestTree
+testPropsInHaskell =
+  testGroup
+    "Haskell"
+    [ test_Hask_MintValueBuiltinData
+    , test_Hask_AssetClassIsEitherMintedOrBurned
+    , test_Hask_MintValueMintedIsPositive
+    , test_Hask_MintValueBurnedIsPositive
+    ]
+
+test_Hask_MintValueBuiltinData :: TestTree
+test_Hask_MintValueBuiltinData =
+  testProperty "Builtin data representation of MintValue and Value is the same"
+    $ scaleTestsBy 5 prop_MintValueBuiltinData
+
+test_Hask_AssetClassIsEitherMintedOrBurned :: TestTree
+test_Hask_AssetClassIsEitherMintedOrBurned =
+  testProperty "Asset class is either minted or burned"
+    $ scaleTestsBy 5 prop_AssetClassIsEitherMintedOrBurned
+
+test_Hask_MintValueMintedIsPositive :: TestTree
+test_Hask_MintValueMintedIsPositive =
+  testProperty "MintValue minted is positive"
+    $ scaleTestsBy 5 prop_MintValueMintedIsPositive
+
+test_Hask_MintValueBurnedIsPositive :: TestTree
+test_Hask_MintValueBurnedIsPositive =
+  testProperty "MintValue burned is positive"
+    $ scaleTestsBy 5 prop_MintValueBurnedIsPositive
+
+--------------------------------------------------------------------------------
+-- CEK-evaluated properties ----------------------------------------------------
+
+testPropsInPlinth :: TestTree
+testPropsInPlinth =
+  testGroup
+    "Plinth"
+    [ test_Plinth_MintValueBuiltinData
+    , test_Plinth_AssetClassIsEitherMintedOrBurned
+    , test_Plinth_MintValueMintedIsPositive
+    , test_Plinth_MintValueBurnedIsPositive
+    ]
+
+test_Plinth_MintValueBuiltinData :: TestTree
+test_Plinth_MintValueBuiltinData =
+  testProperty "Builtin data representation of MintValue and Value is the same"
+    $ cekProp
+    . \values ->
```

Comment:
Does this use some `Arbitrary` instance for `Value`?

Reply:
Yes, it imports existing `instance Arbitrary Value`